lgc/builder/InOutBuilder.cpp

```diff
@@ -65,6 +65,162 @@ Value *InOutBuilder::CreateReadGenericInput(Type *resultTy, unsigned location, V
+                                              unsigned locationCount, InOutInfo inputInfo, Value *vertexIndex,
+                                              const Twine &instName) {
+  assert(!resultTy->isAggregateType());
+  assert(inputInfo.getInterpMode() == InOutInfo::InterpModeCustom);
+  assert(m_shaderStage == ShaderStageFragment);
+
+  // Fold constant pLocationOffset into location.
+  if (auto constLocOffset = dyn_cast<ConstantInt>(locationOffset)) {
+    location += constLocOffset->getZExtValue();
+    locationOffset = getInt32(0);
+    locationCount = (resultTy->getPrimitiveSizeInBits() + 127U) / 128U;
+  }
+
+  // Mark the usage of the input/output.
+  markGenericInputOutputUsage(false, location, locationCount, inputInfo, vertexIndex);
+
+  // Fixing the order
+  bool parity = false;
+  Value *odd = nullptr;
+  Value *even = nullptr;
+
+  auto primType = getPipelineState()->getPrimitiveType();
+  switch (primType) {
+  case PrimitiveType::Point:
+    break;
+  case PrimitiveType::Line_List:
+  case PrimitiveType::Line_Strip:
+    break;
+  case PrimitiveType::Triangle_List: {
+    switch (dyn_cast<ConstantInt>(vertexIndex)->getZExtValue()) {
+    case 0: {
+      vertexIndex = getInt32(2);
+      break;
+    }
+    case 1: {
+      vertexIndex = getInt32(0);
+      break;
+    }
+    case 2: {
+      vertexIndex = getInt32(1);
+      break;
+    }
+    }
+    break;
+  }
+  case PrimitiveType::Triangle_Strip:
+  case PrimitiveType::Triangle_Fan:
+  case PrimitiveType::Triangle_Strip_Adjacency: {
+    switch (dyn_cast<ConstantInt>(vertexIndex)->getZExtValue()) {
+    case 0: {
+      odd = getInt32(2);
+      even = vertexIndex;
+      break;
+    }
+    case 1: {
+      odd = getInt32(0);
+      even = vertexIndex;
+      break;
+    }
+    case 2: {
+      odd = getInt32(1);
+      even = vertexIndex;
+      break;
+    }
+    }
+    parity = true;
+    break;
+  }
+  case PrimitiveType::Triangle_List_Adjacency: {
+    switch (dyn_cast<ConstantInt>(vertexIndex)->getZExtValue()) {
+    case 0: {
+      odd = getInt32(1);
+      even = vertexIndex;
+      break;
+    }
+    case 1: {
+      odd = getInt32(2);
+      even = vertexIndex;
+      break;
+    }
+    case 2: {
+      odd = getInt32(0);
+      even = vertexIndex;
+      break;
+    }
+    }
+    parity = true;
+    break;
+  }
+  default:
+    llvm_unreachable("Unable to get vertices per primitive!");
+    break;
+  }
+
+  Value *result = nullptr;
+  if (parity) {
+    // Odd
+    SmallVector<Value *, 6> args;
+    args.push_back(getInt32(location));
+    args.push_back(locationOffset);
+    args.push_back(elemIdx);
+    args.push_back(getInt32(inputInfo.getInterpMode()));
+    args.push_back(odd);
+    std::string callName(lgcName::InputImportInterpolant);
+    addTypeMangling(resultTy, args, callName);
+    auto oddRes = emitCall(callName, resultTy, args, {Attribute::ReadOnly, Attribute::WillReturn}, &*GetInsertPoint());
```

Comment:
Do you think it makes sense to move this code into a lambda instead of duplicating it?
I.e.
```c++

auto readInput = [&](Value *vertexIndex) {
  std::string callName = lgcName::InputImportInterpolant;
  SmallVector<Value *, 5> args({
    getInt32(location),
    locationOffset,
    elemIdx,
    getInt32(inputInfo.getInterpMode()),
    vertexIndex,
  });
  addTypeMangling(resultTy, args, callName);
  return emitCall(callName, resultTy, args, {Attribute::ReadOnly, Attribute::WillReturn}, &*GetInsertPoint());
}

if (parity) {
  auto oddRes = readInput(odd);
  auto evenRes = readInput(even);
  …
} else {
  result = readInput(vertexIndex);
}
```

Reply:
Yes, it is clearer. Thanks.